src/client/wrap.cpp

```diff
@@ -103,17 +105,140 @@ GeneratedModel::GeneratedModel(const size_t& initShapeIdx, const std::vector<dou
 
 namespace {
 
-void extractMainShapeAttributes(const py::dict& shapeAttr, std::wstring& ruleFile, std::wstring& startRule,
-                                int32_t& seed, std::wstring& shapeName, pcu::AttributeMapPtr& convertShapeAttr) {
+bool getResolveMap(const std::string& rulePackagePath, pcu::ResolveMapPtr* resolveMap) {
+	if (!rulePackagePath.empty()) {
+		LOG_INF << "using rule package " << rulePackagePath << std::endl;
+
+		const std::string u8rpkURI = pcu::toFileURI(rulePackagePath);
+		prt::Status status = prt::STATUS_UNSPECIFIED_ERROR;
+		try {
+			resolveMap->reset(prt::createResolveMap(pcu::toUTF16FromUTF8(u8rpkURI).c_str(), nullptr, &status));
+		}
+		catch (const std::exception& e) {
+			pybind11::print("CAUGHT EXCEPTION:", e.what());
+			return false;
+		}
+
+		if (resolveMap && (status == prt::STATUS_OK)) {
+			LOG_DBG << "resolve map = " << pcu::objectToXML(resolveMap->get()) << std::endl;
+		}
+		else {
+			LOG_ERR << "getting resolve map from '" << rulePackagePath << "' failed, aborting.";
+			return false;
+		}
+	}
+
+	return true;
+}
+
+std::wstring getRuleFileEntry(const prt::ResolveMap* resolveMap) {
+	const std::wstring sCGB(L".cgb");
+
+	size_t nKeys;
+	wchar_t const* const* keys = resolveMap->getKeys(&nKeys);
+	for (size_t k = 0; k < nKeys; k++) {
+		const std::wstring key(keys[k]);

```

Comment:
constructing a std::wstring could be avoided by directly using the c-string functions from https://en.cppreference.com/w/cpp/header/cwchar (string examination)